lib/core/src/display/navigation/navigator.rs

```diff
@@ -0,0 +1,158 @@
+        }
+        self.animate(camera, delta_seconds);
+    }
+
+    /// For zooming into a desired focused position.
+    fn zoom(&mut self, camera:&mut Camera, zoom:ZoomEvent) {
+        if let CameraType::Perspective(persp) = camera.camera_type() {
+            let point        = zoom.focus - self.dom.position();
+            // 0..1 normalization
+            let normalized_x = point.x / self.dom.dimensions().x;
+            let normalized_y = point.y / self.dom.dimensions().y;
+            // -1..1 normalization
+            let normalized_x = (normalized_x - 0.5) * 2.0;
+            let normalized_y = (normalized_y - 0.5) * 2.0;
+
+            let z =  zoom.amount * 6.0;
+            let x = -normalized_x * z / persp.aspect;
+            let y =  normalized_y * z / camera.get_y_scale();
+
+            let animation = &mut self.anim;
+            animation.desired_pos += Vector3::new(x, y, z);
+            if  animation.desired_pos.z < persp.near + 1.0 {
+                animation.desired_pos.z = persp.near + 1.0;
+            }
+        }
+    }
+
+    /// To pan to an offset.
+    fn pan(&mut self, camera:&mut Camera, pan:PanEvent) {
+        if let CameraType::Perspective(_) = camera.camera_type() {
+            let base_z = self.dom.dimensions().y / 2.0 * camera.get_y_scale();
+            let scale  = camera.position().z / base_z;
+
+            let x = pan.movement.x * scale;
+            let y = pan.movement.y * scale;
+            let z = 0.0;
+
+            self.anim.desired_pos += Vector3::new(x, y, z);
+        }
+    }
+
+    /// Animates the camera in a fixed frame rate.
+    fn animate(&mut self, camera:&mut Camera, dt:f32) {
+        self.anim.time_count += dt;
+        let frame_time = 1.0 / self.anim.desired_rate;
+        while self.anim.time_count > frame_time {
+            self.anim.time_count  -= frame_time;
+            self.tick(camera);
+        }
+    }
+
+    /// Runs camera animation.
+    fn tick(&mut self, camera:&mut Camera) {
+        let animation     = &mut self.anim;
+        let desired_pos   = animation.desired_pos;
+        let cam_delta     = desired_pos - camera.position();
+        let cam_delta_len = cam_delta.magnitude();
+
+        if cam_delta_len > 0.0 {
+            let force_val     = cam_delta_len * animation.spring_coeff;
+            let force         = cam_delta.normalize() * force_val;
+            let acc           = force / animation.mass;
+            animation.vel    += acc;
+            let new_vel_val   = animation.vel.magnitude();
+
+            if  new_vel_val   < animation.min_vel {
+                animation.vel          = Vector3::new(0.0, 0.0, 0.0);
+                *camera.position_mut() = animation.desired_pos;
+            } else {
+                animation.vel = animation.vel.normalize() * new_vel_val;
+                let position  = camera.position_mut();
+                *position    += animation.vel;
+            }
+
+            if new_vel_val != 0.0 {
```

Comment:
Again, what is the purpose of this code? The condition should be refactored to a variable with a name telling about the purpose. Or the code may be as well refactored to well named function.

Reply:
Answer in the comment bellow